lib/Runtime/HostManager/HostManager.cpp

```diff
@@ -0,0 +1,159 @@
+/**
+ * Copyright (c) 2017-present, Facebook, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "glow/Runtime/HostManager/HostManager.h"
+#include "glow/Backends/DeviceManager.h"
+#include "glow/Graph/Context.h"
+#include "glow/Partitioner/Partitioner.h"
+#include "glow/Runtime/Executor/Executor.h"
+#include "glow/Runtime/Provisioner/Provisioner.h"
+#include "glow/Runtime/RuntimeTypes.h"
+
+#include <future>
+#include <queue>
+
+using namespace glow;
+using namespace runtime;
+
+HostManager::HostManager(const std::vector<DeviceConfig> &configs) {
+  DeviceIDTy deviceCount = 0;
+  provisioner_.reset(new Provisioner());
+  if (configs.size() > 0) {
+    backend_.reset(createBackend(configs[0].backendKind));
+  }
+  for (auto &config : configs) {
+    auto newDevice = DeviceManager::createDeviceManager(config.backendKind,
+                                                        config.deviceName);
+    devices_.emplace(deviceCount, newDevice);
+    deviceCount++;
+  }
+
+  executor_.reset(createExecutor(devices_));
+}
+
+HostManager::~HostManager() { clearHost(); }
+
+ResultCode HostManager::addNetwork(Module *M) {
+  std::lock_guard<std::mutex> networkLock(networkLock_);
+  std::vector<DeviceInfo> deviceInfo;
+  for (auto &device : devices_) {
+    DeviceInfo info = DeviceInfo();
+    info.availableMemory = device.second->getAvailableMemory();
+    deviceInfo.push_back(info);
+  }
+  // Optimize functions before passing to partitioner.
+  // Currently hardcoding inference.
+  if (backend_) {
+    for (auto F : M->getFunctions()) {
+      backend_->optimizeFunction(CompilationMode::Infer, F);
+    }
+  }
+  auto partitioner = Partitioner(M, deviceInfo);
+  // auto provisioner = llvm::make_unique<Provisioner>();
+  auto nodeList = std::move(partitioner.Partition());
+  auto result = provisioner_->provision(nodeList.roots, devices_, *M);
+
+  for (auto &node : nodeList.roots) {
+    roots_.emplace(node->name, std::move(node));
+  }
+  for (auto &node : nodeList.nodes) {
+    networks_.emplace(node->name, std::move(node));
+  }
+  return result;
+}
+
+void HostManager::removeNetwork(llvm::StringRef networkName) {
```

Comment:
what if active requests are in flight for the network being removed?

Reply:
A great question. That's related to #2297 which will need to get tackled.